pygpu/collectives.pyx

```diff
@@ -33,6 +36,40 @@ cdef class GpuCommCliqueId:
         if comm_id is not None:
             self.comm_id = comm_id
 
+    def __getbuffer__(self, Py_buffer* buffer, int flags):
+        if buffer == NULL:
+            raise BufferError, "NULL buffer view in getbuffer"
+
+        buffer.buf = <char*>self.c_comm_id.internal
+        buffer.obj = self
+        buffer.len = GA_COMM_ID_BYTES * sizeof(char)
+        buffer.readonly = 0
+        buffer.itemsize = sizeof(char)
+        if flags & PyBUF_FORMAT == PyBUF_FORMAT:
+            buffer.format = 'b'
+        else:
+            buffer.format = NULL
+        buffer.ndim = 1
+        if flags & PyBUF_ND == PyBUF_ND:
+            buffer.shape = <Py_ssize_t*>calloc(1, sizeof(Py_ssize_t))
+            buffer.shape[0] = GA_COMM_ID_BYTES
+        else:
+            buffer.shape = NULL
+        if flags & PyBUF_STRIDES == PyBUF_STRIDES:
+            buffer.strides = &buffer.itemsize
+        else:
+            buffer.strides = NULL
+        buffer.suboffsets = NULL
+        buffer.internal = NULL
+        Py_INCREF(self)
+
+    def __releasebuffer__(self, Py_buffer* buffer):
+        if buffer == NULL:
+            raise BufferError, "NULL buffer view in releasebuffer"
+        if buffer.shape != NULL:
+            free(buffer.shape)
+        Py_DECREF(self)
```

Comment:
Yes you are right I was thinking about this. I want to expose a buffer
interface so that we can avoid the cost of creating and copying bytestrings
multiple times. Certainly the cost of this isn't that great, but I felt
better avoiding this should we had a way to do this.

I suppose that the common id will be shared among processes through mpi. In
platoon we can use mpi4py which supports objects that expose a buffer
interface as inputs. This will have zero copies other than those that mpi
internally does.

If this benefit is considered less than the cost of having more complex
code, then this will be removed. What do you say??
On Jul 8, 2016 19:28, "abergeron" <notifications@github.com> wrote:

> In pygpu/collectives.pyx
> <https://github.com/Theano/libgpuarray/pull/206#discussion_r70102010>:
>
> > +        else:
> > +            buffer.shape = NULL
> > +        if flags & PyBUF_STRIDES == PyBUF_STRIDES:
> > +            buffer.strides = &buffer.itemsize
> > +        else:
> > +            buffer.strides = NULL
> > +        buffer.suboffsets = NULL
> > +        buffer.internal = NULL
> > +        Py_INCREF(self)
> > +
> > +    def __releasebuffer__(self, Py_buffer* buffer):
> > +        if buffer == NULL:
> > +            raise BufferError, "NULL buffer view in releasebuffer"
> > +        if buffer.shape != NULL:
> > +            free(buffer.shape)
> > +        Py_DECREF(self)
>
> Does this really add something that we want or need? I don't think that
> there will be people doing a lot of manipulation on the internal of the
> clique ID.
>
> And if that is not useful I would remove it since it just makes for more
> code to maintain.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/libgpuarray/pull/206/files/8fd4fc1115075e2e340020abb69336478482034b..dc42c5388270fadf10c9bc60588cb44e580c5825#r70102010>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AD5VATeNS99sKwN83kZN_FydiLYgXtvRks5qTnqigaJpZM4JAkNl>
> .
>